File: scripts/geocode_regions.mjs
#!/usr/bin/env node
/**
 * geocode_regions.mjs
 *
 * Phase 2C: Fetch boundaries for remaining regions without geographic data
 * using Nominatim (OpenStreetMap) API. Queries admin boundaries (provinces,
 * cantons, prefectures) and wine region locations.
 *
 * For regions that ARE administrative units → polygon boundary (approximate confidence)
 * For wine-concept regions → centroid or small admin boundary (geocoded confidence)
 *
 * Uses data/region_nominatim_queries.json for query overrides.
 * Respects Nominatim rate limit (1 req/sec).
 *
 * Usage:
 *   node scripts/geocode_regions.mjs --dry-run   # Preview queries (default)
 *   node scripts/geocode_regions.mjs --apply      # Actually geocode and insert
 */

import { readFileSync } from 'fs';
import { createClient } from '@supabase/supabase-js';

// ── Load .env ───────────────────────────────────────────────
const envPath = new URL('../.env', import.meta.url).pathname.replace(/^\/([A-Z]:)/, '$1');
const envContent = readFileSync(envPath, 'utf-8');
for (const line of envContent.split('\n')) {
  const trimmed = line.replace(/\r/g, '').trim();
  if (!trimmed || trimmed.startsWith('#')) continue;
  const eqIdx = trimmed.indexOf('=');
  if (eqIdx === -1) continue;
  const key = trimmed.slice(0, eqIdx);
  const val = trimmed.slice(eqIdx + 1);
  if (!process.env[key]) process.env[key] = val;
}

const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE || process.env.SUPABASE_ANON_KEY;
const sb = createClient(SUPABASE_URL, SUPABASE_KEY);

const APPLY = process.argv.includes('--apply');
const MODE = APPLY ? 'APPLY' : 'DRY RUN';

// ── Load query overrides ────────────────────────────────────
const queriesPath = new URL('../data/region_nominatim_queries.json', import.meta.url)
  .pathname.replace(/^\/([A-Z]:)/, '$1');
const { overrides } = JSON.parse(readFileSync(queriesPath, 'utf8'));

// ── Nominatim helpers ───────────────────────────────────────
const NOMINATIM_URL = 'https://nominatim.openstreetmap.org/search';
const USER_AGENT = 'Loam/1.0 (wine-intelligence-platform)';

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

async function nominatimSearch(query, withPolygon = true) {
  const params = new URLSearchParams({
    q: query,
    format: 'jsonv2',
    limit: '1',
    ...(withPolygon ? { polygon_geojson: '1' } : {}),
  });

  const url = `${NOMINATIM_URL}?${params}`;
  const res = await fetch(url, {
    headers: { 'User-Agent': USER_AGENT },
  });

  if (!res.ok) {
    throw new Error(`Nominatim ${res.status}: ${await res.text()}`);
  }

  const results = await res.json();
  return results.length > 0 ? results[0] : null;
}

/**
 * Simplify GeoJSON geometry to stay under 250KB.
 * Uses coordinate precision rounding + Douglas-Peucker if needed.
 */
function simplifyPrecision(geojson) {
  function roundCoords(coords) {
    if (typeof coords[0] === 'number') {
      return [Math.round(coords[0] * 100000) / 100000, Math.round(coords[1] * 100000) / 100000];
    }
    return coords.map(c => roundCoords(c));
  }
  return { type: geojson.type, coordinates: roundCoords(geojson.coordinates) };
}

function simplifyLine(points, tolerance) {
  if (points.length <= 2) return points;
  let maxDist = 0, maxIdx = 0;
  const [ax, ay] = points[0], [bx, by] = points[points.length - 1];
  const dx = bx - ax, dy = by - ay;
  const lenSq = dx * dx + dy * dy;
  for (let i = 1; i < points.length - 1; i++) {
    let dist;
    if (lenSq === 0) {
      dist = Math.sqrt((points[i][0] - ax) ** 2 + (points[i][1] - ay) ** 2);
    } else {
      const t = Math.max(0, Math.min(1, ((points[i][0] - ax) * dx + (points[i][1] - ay) * dy) / lenSq));
      dist = Math.sqrt((points[i][0] - (ax + t * dx)) ** 2 + (points[i][1] - (ay + t * dy)) ** 2);
    }
    if (dist > maxDist) { maxDist = dist; maxIdx = i; }
  }
  if (maxDist > tolerance) {
    const left = simplifyLine(points.slice(0, maxIdx + 1), tolerance);
    const right = simplifyLine(points.slice(maxIdx), tolerance);
    return left.slice(0, -1).concat(right);
  }
  return [points[0], points[points.length - 1]];
}

function simplifyGeometry(geojson, tolerance) {
  function simplifyRing(ring) {
    const simplified = simplifyLine(ring, tolerance);
    return simplified.length >= 4 ? simplified : ring;
  }
  function simplifyCoords(coords, depth) {
    if (depth === 0) return simplifyRing(coords);
    return coords.map(c => simplifyCoords(c, depth - 1));
  }
  const depth = geojson.type === 'MultiPolygon' ? 2 : geojson.type === 'Polygon' ? 1 : 0;
  if (depth === 0) return geojson;
  return { type: geojson.type, coordinates: simplifyCoords(geojson.coordinates, depth) };
}

function prepareGeometry(geojson) {
  let simplified = simplifyPrecision(geojson);
  let str = JSON.stringify(simplified);

  // Progressive simplification to stay under 250KB
  const tolerances = [0.001, 0.005, 0.01, 0.05];
  for (const tol of tolerances) {
    if (str.length <= 256_000) break;
    simplified = simplifyGeometry(simplified, tol);
    str = JSON.stringify(simplified);
  }

  return { geojson: simplified, sizeKB: (str.length / 1024).toFixed(1) };
}

// ── Main ────────────────────────────────────────────────────
async function main() {
  console.log(`\n=== Geocode Regions — Phase 2C (${MODE}) ===\n`);

  // 1. Get all named regions without boundaries
  const { data: regions, error: regErr } = await sb
    .from('regions')
    .select('id, slug, name, parent_id, country_id, countries!inner(name, iso_code)')
    .is('deleted_at', null)
    .eq('is_catch_all', false)
    .order('name');
  if (regErr) throw regErr;

  // Get existing region boundaries
  const { data: existingBounds, error: boundErr } = await sb
    .from('geographic_boundaries')
    .select('region_id')
    .not('region_id', 'is', null);
  if (boundErr) throw boundErr;

  const hasRegionBoundary = new Set(existingBounds.map(b => b.region_id));
  const remaining = regions.filter(r => !hasRegionBoundary.has(r.id));

  console.log(`Total named regions: ${regions.length}`);
  console.log(`Already have boundaries: ${hasRegionBoundary.size}`);
  console.log(`Remaining to geocode: ${remaining.length}\n`);

  // 2. Process each region
  let polygonCount = 0;
  let centroidCount = 0;
  let skippedCount = 0;
  let errorCount = 0;

  for (const region of remaining) {
    const slug = region.slug;
    const countryName = region.countries.name;
    const override = overrides[slug];
    const level = region.parent_id ? 'L2' : 'L1';

    // Check if should skip
    if (override?.skip) {
      console.log(`  SKIP: ${region.name} (${countryName}) — ${override.reason}`);
      skippedCount++;
      continue;
    }

    // Determine query
    const query = override?.query || `${region.name}, ${countryName}`;

    console.log(`  [${level}] ${region.name} (${countryName})`);
    console.log(`       Query: "${query}"`);

    if (!APPLY) {
      continue;
    }

    // Rate limit
    await sleep(1100);

    try {
      let result = await nominatimSearch(query, true);

      // Try fallback query if no result
      if (!result && override?.fallback) {
        console.log(`       Trying fallback: "${override.fallback}"`);
        await sleep(1100);
        result = await nominatimSearch(override.fallback, true);
      }

      if (!result) {
        console.log(`       ❌ No results found`);
        errorCount++;
        continue;
      }

      const lat = parseFloat(result.lat);
      const lng = parseFloat(result.lon);
      const geojson = result.geojson;
      const osmType = result.osm_type;
      const osmId = result.osm_id;
      const sourceId = `nominatim/${osmType}/${osmId}`;

      // Determine confidence
      const isPolygon = geojson &&
        (geojson.type === 'Polygon' || geojson.type === 'MultiPolygon');
      const confidence = isPolygon ? 'approximate' : 'geocoded';

      if (isPolygon) {
        const { geojson: simplified, sizeKB } = prepareGeometry(geojson);

        console.log(`       ✅ Polygon (${simplified.type}, ${sizeKB} KB) → ${confidence}`);

        const { error: rpcErr } = await sb.rpc('upsert_region_boundary', {
          p_region_id: region.id,
          p_geojson: JSON.stringify(simplified),
          p_source_id: sourceId,
          p_confidence: confidence,
        });

        if (rpcErr) {
          console.log(`       ❌ RPC error: ${rpcErr.message}`);
          errorCount++;
          continue;
        }
        polygonCount++;
      } else {
        console.log(`       📍 Centroid only (${lat.toFixed(4)}, ${lng.toFixed(4)}) → geocoded`);

        const { error: rpcErr } = await sb.rpc('upsert_region_boundary', {
          p_region_id: region.id,
          p_lat: lat,
          p_lng: lng,
          p_source_id: sourceId,
          p_confidence: 'geocoded',
        });

        if (rpcErr) {
          console.log(`       ❌ RPC error: ${rpcErr.message}`);
          errorCount++;
          continue;
        }
        centroidCount++;
      }
    } catch (err) {
      console.log(`       ❌ Error: ${err.message}`);
      errorCount++;
    }
  }

  console.log(`\n=== Summary ===`);
  console.log(`Polygon boundaries: ${polygonCount}`);
  console.log(`Centroid only: ${centroidCount}`);
  console.log(`Skipped: ${skippedCount}`);
  console.log(`Errors: ${errorCount}`);
  console.log(`Total processed: ${polygonCount + centroidCount + skippedCount + errorCount}`);
  console.log(`\nDone!`);
}

main().catch(e => { console.error(e); process.exit(1); });

File: pipeline/geo/geocode_regions.py
"""
Geocodes regions via Nominatim with query overrides from data/region_nominatim_queries.json.
Uses progressive simplification for polygons.

Usage:
  python -m pipeline.geo.geocode_regions --dry-run    # preview (default)
  python -m pipeline.geo.geocode_regions --apply       # apply changes
"""

import sys
import json
import time
import argparse
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

import httpx
from pipeline.lib.db import get_supabase
from pipeline.geo.helpers import (
    nominatim_search, has_polygon, simplify_precision, progressive_simplify,
    RATE_LIMIT_S, fetch_all_paginated,
)

PROJECT_ROOT = Path(__file__).resolve().parents[2]


def main():
    parser = argparse.ArgumentParser(description="Geocode regions via Nominatim")
    parser.add_argument("--apply", action="store_true", help="Apply changes (default is dry-run)")
    parser.add_argument("--dry-run", action="store_true", help="Preview only (default)")
    args = parser.parse_args()

    apply = args.apply
    sb = get_supabase()
    client = httpx.Client(timeout=30)
    mode = "APPLY" if apply else "DRY RUN"
    print(f"\n=== Geocode Regions ({mode}) ===\n")

    # Load query overrides
    overrides_path = PROJECT_ROOT / "data" / "region_nominatim_queries.json"
    overrides = {}
    if overrides_path.exists():
        overrides = json.loads(overrides_path.read_text(encoding="utf-8"))
        print(f"Loaded {len(overrides)} query overrides")

    # Get all regions with country
    regions = fetch_all_paginated(sb, "regions", "id, slug, name, country_id, is_catch_all")

    # Get countries
    countries = sb.table("countries").select("id, name").execute().data
    country_by_id = {c["id"]: c["name"] for c in countries}

    # Get existing boundaries
    existing = fetch_all_paginated(sb, "geographic_boundaries", "id, region_id")
    existing_set = {e["region_id"] for e in existing if e.get("region_id")}

    # Filter to non-catch-all regions without boundaries
    todo = [r for r in regions if not r.get("is_catch_all") and r["id"] not in existing_set]
    print(f"Total regions: {len(regions)}")
    print(f"To geocode: {len(todo)}\n")

    fixed = 0
    failed = 0

    for i, r in enumerate(todo):
        country_name = country_by_id.get(r["country_id"], "Unknown")
        print(f"  [{i + 1}/{len(todo)}] {r['name']} ({country_name})")

        # Build queries
        if r["slug"] in overrides:
            queries = overrides[r["slug"]]
        else:
            queries = [
                f"{r['name']}, {country_name}",
                f"{r['name']} wine region, {country_name}",
            ]

        if not apply:
            print(f"    Would try: {' -> '.join(queries)}")
            continue

        result = None
        for query in queries:
            time.sleep(RATE_LIMIT_S)
            try:
                results = nominatim_search(client, query)
                if results:
                    result = results[0]
                    if has_polygon(result):
                        break
            except Exception as e:
                print(f"    Error: {e}")

        if not result:
            print(f"    FAIL no results")
            failed += 1
            continue

        lat = float(result["lat"])
        lng = float(result["lon"])
        geojson = result.get("geojson")
        osm_type = result.get("osm_type", "")
        osm_id = result.get("osm_id", "")
        source_id = f"nominatim/{osm_type}/{osm_id}"
        is_poly = geojson and geojson.get("type") in ("Polygon", "MultiPolygon")

        if is_poly:
            simplified = progressive_simplify(geojson)
            size_kb = len(json.dumps(simplified)) / 1024
            print(f"    ok Polygon ({simplified['type']}, {size_kb:.1f} KB)")
            sb.rpc("upsert_region_boundary", {
                "p_region_id": r["id"],
                "p_geojson": json.dumps(simplified),
                "p_source_id": source_id,
                "p_confidence": "approximate",
            }).execute()
        else:
            print(f"    ok Centroid ({lat:.4f}, {lng:.4f})")
            sb.rpc("upsert_region_boundary", {
                "p_region_id": r["id"],
                "p_lat": lat, "p_lng": lng,
                "p_source_id": source_id,
                "p_confidence": "geocoded",
            }).execute()
        fixed += 1

    client.close()
    print(f"\n=== Summary ===")
    print(f"Fixed: {fixed}")
    print(f"Failed: {failed}")


if __name__ == "__main__":
    main()
